File: cpp-std/demo21b-condition-variable.cpp
/*
CONDITION VARIABLES
*/


#include <iostream>
#include <thread>
#include <mutex>
#include <condition_variable>
using namespace std;



std::mutex mut;

std::condition_variable conditionVar;

int counter = 0;

constexpr int COUNT_HALT_01 = 3;
constexpr int COUNT_HALT_02 = 6;
constexpr int COUNT_DONE = 10;



// Write numbers 1-3 and 8-10 as permitted by egg()
void foo() {
    for (;;) {
        // Lock mutex and then wait for signal to relase mutex
        std::unique_lock<std::mutex> lk(mut);

        // Wait while egg() operates on counter,
        // Mutex unlocked if condition variable in egg() signaled
        conditionVar.wait(lk);

        ++counter;
        cout << "foo counter = " << counter << endl;

        if (counter >= COUNT_DONE) {
            return;
        }
    }
}



// Write numbers 4-7
void egg() {
    for (;;) {
        std::unique_lock<std::mutex> lk(mut);

        if (counter < COUNT_HALT_01 || counter > COUNT_HALT_02) {
            // Signal to free waiting thread by freeing the mutex
            // Note: foo() is now permitted to modify "counter"
            conditionVar.notify_one();
        }
        else {
            ++counter;
            cout << "egg counter = " << counter << endl;
        }

        if (counter >= COUNT_DONE) {
            return;
        }
    }
}



int main() {
    auto thFoo = std::thread(foo);
    auto thEgg = std::thread(egg);

    thFoo.join();
    thEgg.join();

    return 0;
}

File: python/demo21b_condition_variable.py
'''
CONDITION VARIABLES
'''

import threading



condition_var = threading.Condition()

counter = 0

COUNT_HALT_01 = 3
COUNT_HALT_02 = 6
COUNT_DONE = 10



def foo():
    '''
    Write numbers 1-3 and 8-10 as permitted by egg()
    '''
    global counter

    while True:
        with condition_var:
            condition_var.wait()

            counter += 1
            print(f'foo counter = {counter}')

            if counter >= COUNT_DONE:
                return



def bar():
    '''
    Write numbers 4-7
    '''
    global counter

    while True:
        with condition_var:
            if counter < COUNT_HALT_01 or counter > COUNT_HALT_02:
                # Signal to free waiting thread by freeing the mutex
                # Note: foo() is now permitted to modify "counter"
                condition_var.notify()
            else:
                counter += 1
                print(f'egg counter = {counter}')

            if counter >= COUNT_DONE:
                return



threading.Thread(target=foo).start()
threading.Thread(target=bar).start()
